File: python/triangle/triangle.py
def equilateral(sides):
    return is_triangle(sides) and sides[0] == sides[1] == sides[2]


def isosceles(sides):
    return is_triangle(sides) and sides[0] == sides[1] or sides[0] == sides[2] or sides[1] == sides[2]


def scalene(sides):
    return is_triangle(sides) and not isosceles(sides)


def is_triangle(sides):
    sides.sort()
    return sides[0] > 0 and sides[0] + sides[1] >= sides[2]

File: cpp/triangle/triangle.cpp
#include "triangle.h"
#include <stdexcept>

namespace triangle {
    bool valid(double a, double b, double c) {
        return a > 0 && b > 0 && c > 0 && a + b > c && a + c > b && b + c > a;
    }

    flavor kind(double a, double b, double c) {
        if (!valid(a, b, c)) throw std::domain_error("invalid side lengths");
        else if (a == b && b == c) return equilateral;
        else if (a == b || b == c || a == c) return isosceles;
        else return scalene;
    }
}  // namespace triangle
